pkg/kp/rcstore/consul_store.go

```diff
@@ -171,39 +197,92 @@ func (s *consulStore) Lock(id fields.ID, session string) (bool, error) {
 }
 
 func (s *consulStore) Disable(id fields.ID) error {
-	if err := s.verifyExists(id); err != nil {
-		return err
-	}
-	return s.putOne(id, "disabled", []byte("true"))
+	return s.retryMutate(id, func(rc fields.RC) fields.RC {
+		rc.Disabled = true
+		return rc
+	})
 }
 
 func (s *consulStore) SetDesiredReplicas(id fields.ID, n int) error {
-	if err := s.verifyExists(id); err != nil {
-		return err
-	}
-	return s.putOne(id, "replicas_desired", []byte(strconv.Itoa(n)))
+	return s.retryMutate(id, func(rc fields.RC) fields.RC {
+		rc.ReplicasDesired = n
+		return rc
+	})
 }
 
 func (s *consulStore) Delete(id fields.ID) error {
-	rc, err := s.Get(id)
+	return s.retryMutate(id, nil)
+}
+
+// TODO: this function is almost a verbatim copy of pkg/labels retryMutate, can
+// we find some way to combine them?
+func (s *consulStore) retryMutate(id fields.ID, mutator func(fields.RC) fields.RC) error {
+	err := s.mutateRc(id, mutator)
+	for i := 0; i < s.retries; i++ {
+		if _, ok := err.(CASError); ok {
+			err = s.mutateRc(id, mutator)
+		} else {
+			break
+		}
+	}
+	return err
+}
+
+// performs a safe (ie compare-and-set) mutation of the rc with the given id,
+// using the given function
+// pass nil function to delete
+func (s *consulStore) mutateRc(id fields.ID, mutator func(fields.RC) fields.RC) error {
+	rcp := kp.RCPath(id.String())
+	kvp, meta, err := s.kv.Get(rcp, nil)
 	if err != nil {
 		return err
 	}
-
-	if rc.ReplicasDesired != 0 {
-		return fmt.Errorf("Replication controller %s has %d desired replicas, must be 0 before can be deleted", id, rc.ReplicasDesired)
+	if kvp == nil {
+		return fmt.Errorf("replication controller %s does not exist", id)
 	}
 
-	_, err = s.kv.DeleteTree(kp.RCPath(id.String()), nil)
+	rc, err := s.kvpToRC(kvp)
 	if err != nil {
 		return err
 	}
+	newKVP := &api.KVPair{
+		Key:         rcp,
+		ModifyIndex: meta.LastIndex,
+	}
 
-	// TODO: If this fails, then we have some dangling labels.
-	// Perhaps they can be cleaned up later.
-	return s.forEachLabel(rc, func(id, k, _ string) error {
-		return s.applicator.RemoveLabel(labels.RC, id, k)
-	})
+	var success bool
+	if mutator != nil {
+		newRC := mutator(rc)
+		b, err := json.Marshal(newRC)
+		if err != nil {
+			return err
+		}
+		newKVP.Value = b
+		success, _, err = s.kv.CAS(newKVP, nil)
+	} else {
+		// special logic for the delete case
+		if rc.ReplicasDesired != 0 {
```

Comment:
yeah, I'm probably going to figure out how to write the "block until RC is healthy" function and append that to this PR. However while detecting the state of newly added nodes is easy, detecting that removed nodes are actually gone is a lot less obvious. Still thinking about how to implement.